lib/interceptor/cache.js

```diff
@@ -88,34 +88,72 @@ class CacheHandler extends DecoratorHandler {
 
   onComplete(rawTrailers) {
     if (this.#value) {
+      // get
+      const entries = this.#handler.entries
+      const resHeaders = parseHeaders(this.#value.data.rawHeaders)
+      const reqHeaders = this.#handler.opts
+
+      // set
       this.#value.data.rawTrailers = rawTrailers
-      this.#value.size += rawTrailers?.reduce((xs, x) => xs + x.length, 0) ?? 0
-      this.#store.set(this.#key, this.#value.data, { ttl: this.#value.ttl, size: this.#value.size })
+      this.#value.size = this.#value.size
+        ? this.#value.size + rawTrailers?.reduce((xs, x) => xs + x.length, 0)
+        : 0
+      this.#value.vary = formatVaryData(resHeaders, reqHeaders)
+      entries.push(this.#value)
+      sortEntriesByVary(entries)
+      this.#store.set(this.#key, entries)
     }
     return this.#handler.onComplete(rawTrailers)
   }
 }
 
+function formatVaryData(resHeaders, reqHeaders) {
+  return resHeaders.vary
+    ?.split(',')
+    .map((key) => key.trim().toLowerCase())
+    .map((key) => [key, reqHeaders[key]])
+    .filter(([_key, val]) => val)
+}
+
 // TODO (fix): Async filesystem cache.
 class CacheStore {
   constructor({ maxSize = 1024 * 1024, maxEntrySize = 128 * 1024 }) {
     this.maxSize = maxSize
     this.maxEntrySize = maxEntrySize
-    this.cache = new LRUCache({ maxSize })
+    this.cache = new Map()
   }
 
   set(key, value, opts) {
-    this.cache.set(key, value, opts)
+    this.cache.set(key, value)
   }
 
   get(key) {
     return this.cache.get(key)
   }
 }
 
-function makeKey(opts) {
-  // NOTE: Ignores headers...
-  return `${opts.origin}:${opts.method}:${opts.path}`
+/*
+  Sort entries by number of vary headers in descending order, because
+  we need to compare the most complex response to the request first.
+  A cached response with an empty ´vary´ field will otherwise win every time.
+*/
+function sortEntriesByVary(entries) {
+  entries.sort((a, b) => {
+    const lengthA = a.vary ? a.vary.length : 0
+    const lengthB = b.vary ? b.vary.length : 0
+    return lengthB - lengthA
+  })
+}
```

Comment:
All matching entries should be valid regardless. If there is a response with an empty vary field then that is valid.